### ISSUE TITLE
Connection leak when streaming transaction is cancelled

### ISSUE THREAD

**Issue by jimmydivvy** [tpolecat/doobie#1279]
I've found an issue causing Hikari connections to be left open if a streaming request is cancelled mid-transaction. 

This is happening in production when an HTTP4s request is closed, but I've been able to replicate this with a simple `IO.cancel`

```scala
        f1 <- sql"SELECT 1".query[String].stream.transact(tx).compile.toVector.start
        _ <- IO.sleep(10 millis)
        _ <- f1.cancel
```

In this example, it appears the finaliser fails to run and the connection is leaked. Eventually this results in connection starvation and the service dies. 

I originally found this issue in Doobie 0.7, but I've also replicated it in the current latest version 0.9.2.  I've created a repeatable test case here - https://github.com/jimmydivvy/DoobieLeakTest

Possible related to #879?

**Comment by jatcwang** [comment]
Thanks for the repro repo. I can reproduce this on `GraalVM CE 20.2.0` too. Upgraded to fs2 2.4.4 doesn't help so I'll keep digging.

**Comment by jatcwang** [comment]
This seems related to https://github.com/typelevel/fs2/issues/1140, though it should work

**Comment by jimmydivvy** [comment]
> This seems related to [typelevel/fs2#1140](https://github.com/typelevel/fs2/issues/1140), though it should work

I wrote a similar test case with just FS2 by itself, and it looks like IO.cancel does correctly calls the FS2 finalisers now.

**Comment by jatcwang** [comment]
@jimmydivvy try one with nested Stream.resource calls, as that's what https://github.com/tpolecat/doobie/blob/3e29e35f249d8efa758a5c3b2da5f6f5601e2921/modules/core/src/main/scala/doobie/util/transactor.scala#L174 is doing.

I added some printlns, and I can see that that Connection is acquired, but it never got to calling `preparedStatement` on it (which it would if you give it more time in LeakTest.scala)

**Comment by jimmydivvy** [comment]
Nested Stream.resource calls worked fine. I wonder if the use of Free and mapK/translate in part of the stream messes things up? I might see if I can replicate that in a smaller scale example.

**Comment by Daenyth** [comment]
> I wonder if the use of Free and mapK/translate in part of the stream messes things up?

Likely; `translate` can't connect cancellation - it needs to be a [`translateInterruptable`](https://oss.sonatype.org/service/local/repositories/releases/archive/co/fs2/fs2-core_2.12/2.4.4/fs2-core_2.12-2.4.4-javadoc.jar/!/fs2/Stream.html#translateInterruptible[F2[x]%3E:F[x],G[_]](u:F2~%3EG)(implicitevidence$53:cats.effect.Concurrent[G]):fs2.Stream[G,O])

**Comment by jatcwang** [comment]
`translateInterruptable` seems like something we need to fix too, but the problem persists even after I've made the change :/

**Comment by jatcwang** [comment]
Quick update: This resource leak should be fixed in the next version of fs2 (2.4.5?). https://github.com/typelevel/fs2/pull/2073 https://github.com/typelevel/fs2/pull/2074

I'll need to do some more testing around using `translateInterruptable`, but I think all it brings is allowing us to interrupt the execution of the `Stream` created from interpreting ConnectionIO (arguably, this is the expensive part of the stream we DO want to cancel). It'll be a binary incompatible change due to requiring `Concurrent[F]`

**Comment by jatcwang** [comment]
@jimmydivvy Can this be closed? Latest fs2 should have fixed this issue
